--- Python-2.7.12/Lib/distutils/unixccompiler.py.org	2016-10-01 23:22:22.896883982 +0000
+++ Python-2.7.12/Lib/distutils/unixccompiler.py	2016-10-01 23:24:56.072048595 +0000
@@ -120,6 +120,10 @@
         try:
             self.spawn(compiler_so + cc_args + [src, '-o', obj] +
                        extra_postargs)
+            self.spawn(compiler_so + cc_args+ ["-march=haswell", "-O3", "-fno-semantic-interposition", "-ffat-lto-objects", "-flto=4"] + [src, '-o', obj + ".avx2"] +
+                       extra_postargs)                      
+            self.spawn(compiler_so + cc_args+ ["-march=skylake-avx512", "-O3", "-fno-semantic-interposition", "-ffat-lto-objects", "-flto=4"] + [src, '-o', obj + ".avx512"] +
+                       extra_postargs)                      
         except DistutilsExecError, msg:
             raise CompileError, msg
 
@@ -198,6 +200,24 @@
                     linker = _osx_support.compiler_fixup(linker, ld_args)
 
                 self.spawn(linker + ld_args)
+                ld_args = ([obj + ".avx2" for obj in objects] + self.objects +
+                       lib_opts + ['-o', output_filename + ".avx2"])
+                if debug:
+                    ld_args[:0] = ['-g']
+                if extra_preargs:
+                    ld_args[:0] = extra_preargs
+                if extra_postargs:
+                    ld_args.extend(extra_postargs)
+                self.spawn(linker + ld_args)
+                ld_args = ([obj + ".avx512" for obj in objects] + self.objects +
+                       lib_opts + ['-o', output_filename + ".avx512"])
+                if debug:
+                    ld_args[:0] = ['-g']
+                if extra_preargs:
+                    ld_args[:0] = extra_preargs
+                if extra_postargs:
+                    ld_args.extend(extra_postargs)
+                self.spawn(linker + ld_args)
             except DistutilsExecError, msg:
                 raise LinkError, msg
         else:
--- Python-2.7.12/Makefile.pre.in~	2016-06-25 21:49:31.000000000 +0000
+++ Python-2.7.12/Makefile.pre.in	2017-07-16 06:26:18.237780631 +0000
@@ -534,6 +534,8 @@
 # This is shared by the math and cmath modules
 Modules/_math.o: Modules/_math.c Modules/_math.h
 	$(CC) -c $(CCSHARED) $(PY_CFLAGS) -o $@ $<
+	$(CC) -c $(CCSHARED) $(PY_CFLAGS) -march=haswell -o $@.avx2 $<
+
 
 # Build the shared modules
 # Under GNU make, MAKEFLAGS are sorted and normalized; the 's' for
